### PR TITLE
Notes on Alpino support

### PR DESCRIPTION
**Pull request by LourensVeen**
Notes on Alpino support, also see inline comments. We have a pretty big issue here because the older build we were using seems to be no longer available, and the newer ones don't work with xtas. We may have to ask upstream for modifications to keep Alpino and xtas working together.

### CODE REVIEW

File: xtas/tasks/_alpino.py [hunk 1]
```diff
@@ -33,7 +33,7 @@
 
 log = logging.getLogger(__name__)
 
-CMD_PARSE = ["bin/Alpino", "end_hook=dependencies", "-parse"]
```
Comment (LourensVeen) [inline xtas/tasks/_alpino.py]: The older version linked above seems to be gone, and the new ones do run, but give different results, which I'm not sure are usable for us.

File: xtas/tasks/_alpino.py [hunk 2]
```diff
@@ -78,7 +78,7 @@ def parse_raw(tokens):
     return parse
 
 
-def interpret_parse(parse):
```
Comment (LourensVeen) [inline xtas/tasks/_alpino.py]: I don't see any easy simplifications, but this function is pretty difficult to read.

File: xtas/tasks/_alpino.py [hunk 3]
```diff
@@ -113,7 +113,7 @@ def get_dep(line):
                 tokens=list(tokens.values()))
 
 
-def interpret_token(lemma, word, begin, _end, major_pos, _pos2, pos):
```
Comment (LourensVeen) [inline xtas/tasks/_alpino.py]: Here we have major_pos, _pos2, pos, and major, which gets rather confusing. Maybe this is just Alpino terminology and we can't do much about it, but if we can, we should I think.

File: xtas/tasks/_alpino.py [hunk 4]
```diff
@@ -175,7 +175,7 @@ def interpret_token(lemma, word, begin, _end, major_pos, _pos2, pos):
           }
 
 
-if __name__ == '__main__':
```
Comment (LourensVeen) [inline xtas/tasks/_alpino.py]: What is this for? Testing?

File: xtas/tasks/_alpino.py (hunk 3 above)
Comment (larsmans) [inline xtas/tasks/_alpino.py]: Alpino produces multiple POS tags per word. I'm not sure if it has any terminology for those, but they need to have some name.

File: xtas/tasks/_alpino.py (hunk 4 above)
Comment (larsmans) [inline xtas/tasks/_alpino.py]: I guess so. I think I'll remove it.

**Comment by larsmans** [comment]
What do you mean exactly by "don't work"? I just downloaded [this version](http://www.let.rug.nl/vannoord/alp/Alpino/binary/binary/Alpino-x86_64-centos5-17570.tar.gz) which produces proper output. I'll try to run the tests.

**Comment by LourensVeen** [comment]
Aha, I've been trying with [http://www.let.rug.nl/~vannoord/alp/Alpino/versions/binary/Alpino-x86_64-Linux-glibc-2.19-20908-sicstus.tar.gz](this one), which is one of these sicstus builds and it fails the test. I guess I never even considered this one, it being so old. I'll see what it does for me.

**Comment by LourensVeen** [comment]
Ouch, dependency error on tcl/tk 8.4. But interestingly, the most recent one depends on 8.5, which I haven't installed either (we're on tcl/tk 8.6 at the moment), and that doesn't complain about anything missing. The good news is that Ubuntu has backwards-compatibility packages, so you can just install tk8.4 and tcl8.4.

Then, it fails test_tokenize() because the last letter seems to fall off, and test_alpino_task() because apparently "Toob is dik" has something to do with Tuesday...

xtas.tests.test_alpino.test_tokenize ... FAIL
xtas.tests.test_alpino.test_parse ... ok
xtas.tests.test_alpino.test_interpret_token ... ok
xtas.tests.test_alpino.test_interpret_parse ... {'lemma': u'ben', 'word': u'is', 'offset': 1, 'pos_minor': u'unacc,sg_heeft,copula', 'pos1': 'V', 'pos': u'verb', 'pos_major': u'verb'}
{'lemma': u'dik', 'word': u'dik', 'offset': 2, 'pos_minor': u'no_e(adv)', 'pos1': 'A', 'pos': u'adj', 'pos_major': u'adjective'}
{'lemma': u'ben', 'word': u'is', 'offset': 1, 'pos_minor': u'unacc,sg_heeft,copula', 'pos1': 'V', 'pos': u'verb', 'pos_major': u'verb'}
{'lemma': u'Toob', 'word': u'Toob', 'offset': 0, 'pos_minor': u'both,PER', 'pos1': 'M', 'pos': u'name', 'pos_major': u'proper_name'}
ok
Test whether the xtas.tasks.single.alpino call works ... FAIL
Test what happens with non-ascii characters in input ... ERROR

======================================================================
ERROR: Test what happens with non-ascii characters in input
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/lourens/NLeSC/projects/xtas/software/LourensVeen-xtas/xtas/tests/test_alpino.py", line 88, in test_alpino_unicode
    assert_equal(tokenize(text).decode("utf-8"), text + "\n")
  File "/home/lourens/Envs/xtas-dev/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xc3 in position 8: invalid continuation byte

======================================================================
FAIL: xtas.tests.test_alpino.test_tokenize
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/lourens/NLeSC/projects/xtas/software/LourensVeen-xtas/xtas/tests/test_alpino.py", line 51, in test_tokenize
    assert_equal(tokenize(text), expected.encode('utf-8'))
AssertionError: "D\xc3\xadt is een zin , met komma nietwaar ?\nEn nog 'n zi\n" != "D\xc3\xadt is een zin , met komma nietwaar ?\nEn nog 'n zin\n"
-------------------- >> begin captured logging << --------------------
xtas.core: INFO: Cannot import xtas_config, falling back to default
xtas.core: INFO: Using Elasticsearch at [{'host': 'localhost', 'port': 9200}]
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: Test whether the xtas.tasks.single.alpino call works
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/lourens/NLeSC/projects/xtas/software/LourensVeen-xtas/xtas/tests/test_alpino.py", line 78, in test_alpino_task
    assert_equal(alpino(_SENT, output='raw').strip(), _PARSE)
AssertionError: 'ben|is|1|2|verb|verb(copula)|verb(unacc,sg_heeft,copula)|hd/predc|dinsdag|di|2|3|noun|noun(tmp)|tmp_noun(de,count,sg)|1\nben|is|1|2|verb|verb(copula)|verb(unacc,sg_heeft,copula)|hd/su|Toob|Toob|0|1|name|name(PER)|proper_name(both,PER)|1' != 'ben|is|1|2|verb|verb(copula)|verb(unacc,sg_heeft,copula)|hd/predc|dik|dik|2|3|adj|adj|adjective(no_e(adv))|1\nben|is|1|2|verb|verb(copula)|verb(unacc,sg_heeft,copula)|hd/su|Toob|Toob|0|1|name|name(PER)|proper_name(both,PER)|1'

----------------------------------------------------------------------
Ran 6 tests in 7.464s

FAILED (errors=1, failures=2)

**Comment by larsmans** [comment]
That version is actually old. It has a broken tokenizer that doesn't produce valid UTF-8. Will try a newer version.

**Comment by larsmans** [comment]
> "Toob is dik" has something to do with Tuesday...

That's because the last letter of "dik" is munched off to produce "di", which is interpreted as "dinsdag". Same bug. Workaround is to append a newline before sending data to the tokenizer.

**Comment by larsmans** [comment]
> dependency error on tcl/tk 8.4

I get that too, but not when Alpino is run with `-parse`.

**Comment by larsmans** [comment]
Fixed by b16b28cc8ab39c62380cc6b0fcae8070492ff1db.

**Comment by LourensVeen** [comment]
The link on line 21 is still dead however, what did you really test this with?

**Comment by LourensVeen** [comment]
It works for me now with version Alpino-x86_64-Linux-glibc-2.19-20905-sicstus. Nice work!

**Comment by larsmans** [comment]
Copypaste error, will fix the link!

**Pull request closed without merging**